binders/kafka-binder/spring-cloud-stream-binder-kafka-reactive/src/main/java/org/springframework/cloud/stream/binder/reactorkafka/ReactorKafkaBinder.java

```diff
@@ -0,0 +1,327 @@
+				ExtendedProducerProperties<KafkaProducerProperties>, KafkaTopicProvisioner>
+		implements
+		ExtendedPropertiesBinder<MessageChannel, KafkaConsumerProperties, KafkaProducerProperties> {
+
+	private static final Log log = LogFactory.getLog(ReactorKafkaBinder.class);
+
+	private final KafkaBinderConfigurationProperties configurationProperties;
+
+	private final KafkaExtendedBindingProperties extendedBindingProperties = new KafkaExtendedBindingProperties();
+
+	public ReactorKafkaBinder(KafkaBinderConfigurationProperties configurationProperties,
+			KafkaTopicProvisioner provisioner) {
+
+		super(new String[0], provisioner, null, null);
+		this.configurationProperties = configurationProperties;
+	}
+
+
+	@Override
+	protected MessageHandler createProducerMessageHandler(ProducerDestination destination,
+			ExtendedProducerProperties<KafkaProducerProperties> producerProperties, MessageChannel errorChannel)
+			throws Exception {
+
+		Map<String, Object> configs = createProducerConfigs(producerProperties);
+		// TODO: Move config customizers to core
+//		if (this.producerConfigCustomizer != null) {
+//			this.producerConfigCustomizer.configure(props, bindingNameHolder.get(), destination);
+//			bindingNameHolder.remove();
+//		}
+
+		SenderOptions<Object, Object> opts = SenderOptions.create(configs);
+		// TODO bean for converter.
+		RecordMessageConverter converter = new MessagingMessageConverter();
+		return new ReactorMessageHandler(opts, converter, destination.getName());
+	}
+
+
+	@Override
+	protected MessageProducer createConsumerEndpoint(ConsumerDestination destination, String group,
+			ExtendedConsumerProperties<KafkaConsumerProperties> properties) throws Exception {
+
+		boolean anonymous = !StringUtils.hasText(group);
+		String consumerGroup = anonymous ? "anonymous." + UUID.randomUUID().toString() : group;
+		Map<String, Object> configs = createConsumerConfigs(anonymous, consumerGroup, properties);
+
+		// TODO: Move config customizers to core
+//		if (this.consumerConfigCustomizer != null) {
+//			this.consumerConfigCustomizer.configure(configs, bindingNameHolder.get(), destination);
+//		}
+
+		RecordMessageConverter converter = new MessagingMessageConverter();
+		ReceiverOptions<Object, Object> opts = ReceiverOptions.create(configs)
+			.addAssignListener(parts -> System.out.println("Assigned: " + parts))
+			.subscription(Collections.singletonList(destination.getName()));
+
+		return new MessageProducerSupport() {
+
+			private final KafkaReceiver<Object, Object> receiver = KafkaReceiver.create(opts);
+
+			@SuppressWarnings("unchecked")
+			@Override
+			protected void doStart() {
+				Flux<Message<Object>> flux = receiver
+						.receive()
+						.map(record -> (Message<Object>) converter.toMessage(record, null, null, null));
+				subscribeToPublisher(flux);
+			}
+
+			@Override
+			protected void doStop() {
+				// TODO how do we dispose of the Disposable created by subscribeToPublisher?
```

Comment:
@artembilan question for you - it doesn't keep a reference to the `Disposable`.

Reply:
It does it by itself:
```Flux<? extends Message<?>> messageFlux =
				Flux.from(publisher)
						.map(this::trackMessageIfAny)
						.doOnComplete(this::stop)
						.doOnCancel(this::stop)
						.takeWhile((message) -> isActive());
```

That `TakeWhileSubscriber` does this:
```
if (!b) {
	s.cancel();

	onComplete();

	return;
}
```
So, nothing to dispose.

Reply:
Thanks.

Reply:
@artembilan There is still a problem here; there is nothing to signal the flux completion.

See https://github.com/spring-cloud/spring-cloud-stream/blob/0f213fedf67d9269e7f984a806d00c5156ac3780/binders/kafka-binder/spring-cloud-stream-binder-kafka-reactive/src/test/java/org/springframework/cloud/stream/binder/reactorkafka/ReactorKafkaBinderTests.java#L109-L110

If you put a breakpoint on the unbind and step over, you see that the `"reactive-kafka-foo-1"` thread is still running.

We normally shutdown the consumer by calling `dispose()` on the `Disposable` returned by the `subscribe()`...

https://github.com/spring-cloud/spring-cloud-stream/blob/dcaa621b3c72bfcce2423622a0e272513cb463c5/binders/kafka-binder/spring-cloud-stream-binder-kafka-reactive/src/main/java/org/springframework/cloud/stream/binder/reactorkafka/ReactorKafkaBinder.java#L140

I think we need a hook to be able to explicitly perform the `dispose()` when the binding is stopped or unbound; perhaps that should be done in `super.doStop()` ?

Reply:
That `dispose()` is called in the `FluxMessageChannel.destroy()`.

I'm not sure why `KafkaConsumer` is not stopped when a `cancel()` is called for its `Flux`.
Perhaps that's something has to be improved in `reactor-kafka`?

I'll try to debug your test later.

Thanks

Reply:
II might be wrong but I don’t think unbinding a binding destroys the channel. Stopping the binding certainly doesn’t, but that should cancel the disposable.

Reply:
I have figured out the problem, Gary.

So, that `TakeWhileSubscriber` does an action we need only when there is a data streaming from a `Publisher`.
In your test we have only one record published to Kafka. And then we just go to the `stop()` immediately.

We really need an additional hook to unsubscribe from the source `Publsher` when endpoint is stopped externally.
It would be great if we would have something what would signal the subscriber to cancel...

Not sure which `Flux` operator to use here...

```
.doOnSubscribe(subs -> this.subscription = subs);
```
?

Reply:
Thanks; that works - https://github.com/spring-cloud/spring-cloud-stream/pull/2295